File: pbhhg_py/interpret.py
import abc
import dataclasses
from typing import Generator, Generic, Sequence, TypeVar

from pbhhg_py import abstract_syntax as AS
from pbhhg_py import builtins
from pbhhg_py import error
from pbhhg_py import parse
from pbhhg_py import utils

_T = TypeVar("_T")


@dataclasses.dataclass
class ComputationRequest:
    value: AS.Expr


@dataclasses.dataclass
class ComputationResponse:
    value: AS.StrictValue | AS.UnsuspectedHangeulError


@dataclasses.dataclass
class ComputationResult(Generic[_T]):
    value: _T | AS.UnsuspectedHangeulError


class StackFrameBase(Generic[_T]):
    def __init__(self, coroutine: Generator[AS.Value, AS.StrictValue, _T]):
        self._coroutine = coroutine

    def communicate(
        self,
        response: ComputationResponse | None,
    ) -> ComputationRequest | ComputationResult[_T]:
        try:
            while True:
                if response is None:
                    request = next(self._coroutine)
                elif isinstance(response.value, AS.UnsuspectedHangeulError):
                    request = self._coroutine.throw(response.value)
                else:
                    request = self._coroutine.send(response.value)

                if isinstance(request, AS.Expr):
                    if not request.cache_box.value:
                        return ComputationRequest(request)
                    request = request.cache_box.value
                response = ComputationResponse(request)

        except StopIteration as result:
            return ComputationResult(result.value)
        except AS.UnsuspectedHangeulError as err:
            return ComputationResult(err)


class StackFrame(StackFrameBase[AS.Value]):
    def __init__(self, request: ComputationRequest):
        self._cache_box = request.value.cache_box
        super().__init__(interpret(request.value))

    def communicate(
        self, response: ComputationResponse | None
    ) -> ComputationRequest | ComputationResult[AS.Value]:
        request_or_result = super().communicate(response)
        if isinstance(request_or_result, ComputationResult):
            if isinstance(request_or_result.value, AS.Expr):
                request_or_result.value.cache_box.requestor = self._cache_box
            else:
                self._cache_box.resolve(request_or_result.value)
        return request_or_result


class DebuggerBase(abc.ABC):
    @abc.abstractmethod
    def before_eval(self, depth: int, expr: AS.Expr) -> None:
        del depth, expr  # Unused
        raise NotImplementedError

    @abc.abstractmethod
    def after_eval(
        self,
        depth: int,
        expr: AS.Expr,
        result: AS.StrictValue | AS.UnsuspectedHangeulError,
    ) -> None:
        del depth, expr, result  # Unused
        raise NotImplementedError


def evaluate(
    coroutine: Generator[AS.Value, AS.StrictValue, _T],
    debugger: DebuggerBase | None = None,
) -> _T:
    """Executes coroutine and provides strictly evaluated values
    as it requests."""
    MAX_STACK_SIZE = 5000
    head = StackFrameBase(coroutine)
    tail: list[StackFrame] = []
    debug_stack: list[list[AS.Expr]] = []  # same index as tail
    depth = 0

    response: ComputationResponse | None = None
    while True:
        if tail:
            if len(tail) >= MAX_STACK_SIZE:
                raise RuntimeError("Maximum Stack Size Exceeded.")

            request_or_result = tail[-1].communicate(response)
            if isinstance(request_or_result, ComputationRequest):
                tail.append(StackFrame(request_or_result))
                response = None

                if debugger:
                    depth += 1
                    debug_stack.append([request_or_result.value])
                    debugger.before_eval(depth, request_or_result.value)

            elif isinstance(request_or_result.value, AS.Expr):
                request = ComputationRequest(request_or_result.value)
                tail[-1] = StackFrame(request)
                response = None

                if debugger:
                    depth += 1
                    debug_stack[-1].append(request_or_result.value)
                    debugger.before_eval(depth, request_or_result.value)

            else:
                tail.pop()
                response = ComputationResponse(request_or_result.value)

                if debugger:
                    waiting_exprs = debug_stack.pop()
                    for expr in reversed(waiting_exprs):
                        debugger.after_eval(
                            depth, expr, request_or_result.value
                        )
                        depth -= 1
        else:
            request_or_result = head.communicate(response)
            if isinstance(request_or_result, ComputationRequest):
                tail.append(StackFrame(request_or_result))
                response = None

                if debugger:
                    depth += 1
                    debug_stack.append([request_or_result.value])
                    debugger.before_eval(depth, request_or_result.value)

            else:  # all finished
                result = request_or_result.value
                if isinstance(result, AS.UnsuspectedHangeulError):
                    # TODO(dragonteros): handle debugging at outer-most frame
                    raise result
                return result


def interpret(value: AS.Expr) -> AS.EvalContext:
    """Interpreter coroutine.

    Args:
        value: an Expr value to interpret

    Yields:
        expr: an Expr value to send to main routine
            which in turn sends strictly evaluated value

    Returns:
        A little more evaluated version of value.
    """
    expr = value.expr
    env = value.env

    if value.cache_box.value:
        cache = value.cache_box.value
        if isinstance(cache, AS.UnsuspectedHangeulError):
            raise cache
        return cache

    if isinstance(expr, AS.Literal):
        return AS.Integer(expr.value)

    elif isinstance(expr, AS.FunRef):
        try:
            return env.funs[-expr.rel - 1]
        except IndexError:
            raise error.UnsuspectedHangeulOutOfRangeError(
                expr.metadata, "함수 참조의 범위를 벗어났습니다."
            )

    elif isinstance(expr, AS.ArgRef):
        assert len(env.funs) == len(env.args)
        relA, relF, metadata = expr
        try:
            args = env.args[-relF - 1]
        except IndexError:
            raise error.UnsuspectedHangeulOutOfRangeError(
                metadata, "존재하지 않는 함수에 대한 인수 참조를 시도했습니다."
            )

        relA = yield AS.Expr(relA, env)
        [relA] = utils.check_type(metadata, [relA], AS.Integer)

        if not 0 <= relA.value < len(args):
            raise error.UnsuspectedHangeulOutOfRangeError(
                metadata, f"{relA.value}번째 인수를 참조하는데 {len(args)}개의 인수만 받았습니다."
            )
        return args[relA.value]

    elif isinstance(expr, AS.FunDef):
        funs, args = env
        new_funs = funs[:]
        new_env = AS.Env(new_funs, args)
        closure = AS.Closure(expr.body, new_env)
        new_env.funs.append(closure)
        return closure

    fun, argv, metadata = expr
    fun = AS.Expr(fun, env)
    argv = [AS.Expr(arg, env) for arg in argv]
    fun = yield from utils.strict_functional(metadata, fun)
    recipe = proc_functional(metadata, fun, general_callable=True)
    return (yield from recipe(metadata, argv))


def strict(
    value: AS.Value,
) -> Generator[AS.Value, AS.StrictValue, AS.StrictValue]:
    return (yield value)


def proc_functional(
    metadata: AS.Metadata,
    fun: AS.StrictValue | AS.BuiltinFunction,
    general_callable: bool = False,
) -> AS.Evaluation:
    """
    Args:
        fun: A maybe-Expr value that may correspond to a function.
        general_callable: Allow non-Function callable types.

    Returns:
        A generator that receives argument list and returns maybe-Expr value.
    """
    if isinstance(fun, AS.BuiltinFunction):
        return find_builtin(metadata, fun.literal.value)

    allow = AS.Callable if general_callable else AS.Function
    [fun] = utils.check_type(metadata, [fun], allow)

    if isinstance(fun, AS.Boolean):
        value = fun.value

        def _proc_boolean(
            metadata: AS.Metadata, argv: Sequence[AS.Value]
        ) -> AS.EvalContext:
            utils.check_arity(metadata, argv, 2)
            return argv[0 if value else 1]
            yield

        return _proc_boolean

    elif isinstance(fun, AS.Dict):

        def _proc_dict(
            metadata: AS.Metadata, argv: Sequence[AS.Value]
        ) -> AS.EvalContext:
            utils.check_arity(metadata, argv, 1)
            arg = yield from utils.recursive_strict(argv[0])
            try:
                return fun.value[arg]
            except KeyError:
                raise error.UnsuspectedHangeulNotFoundError(
                    metadata, f"사전에 다음 표제가 없습니다: {arg}"
                ) from None

        return _proc_dict

    elif isinstance(fun, AS.Complex):

        def _proc_complex(
            metadata: AS.Metadata, argv: Sequence[AS.Value]
        ) -> AS.EvalContext:
            [arg] = yield from utils.match_arguments(
                metadata, argv, AS.Integer, 1
            )
            num, idx = fun.value, arg.value
            if idx in [0, 1]:
                return AS.Float(num.real if idx == 0 else num.imag)
            raise error.UnsuspectedHangeulValueError(
                metadata, f"복소수 객체의 {idx}번째 요소를 접근하려고 했습니다."
            )

        return _proc_complex

    elif isinstance(fun, AS.Sequence | AS.ErrorValue):
        seq = fun.value

        def _proc_seq(
            metadata: AS.Metadata, argv: Sequence[AS.Value]
        ) -> AS.EvalContext:
            [arg] = yield from utils.match_arguments(
                metadata, argv, AS.Integer, 1
            )
            idx = arg.value
            if isinstance(seq, tuple):
                try:
                    return seq[idx]
                except IndexError:
                    raise error.UnsuspectedHangeulOutOfRangeError(
                        metadata, "목록의 범위 밖의 번호를 요청했습니다."
                    ) from None
            return utils.guessed_wrap(seq[idx:][:1])

        return _proc_seq

    return fun


def find_builtin(metadata: AS.Metadata, func_id: int):
    """Finds the recipe function corresponding to the builtin function id.

    Args:
        metadata: Source metadata of the caller.
        id: Built-in Function ID

    Returns:
        Corresponding function that takes arguments.
    """
    inst = parse.encode_number(func_id)
    if inst in BUITLINS:
        return BUITLINS[inst]
    raise error.UnsuspectedHangeulTypeError(
        metadata, f"{inst}라는 이름의 기본 제공 함수를 찾지 못했습니다."
    )


def build_builtin_tables():
    table: dict[str, AS.Evaluation] = {}
    for name in builtins.__all__:
        imported = __import__("pbhhg_py.builtins." + name, fromlist=[name])
        new_table = imported.build_tbl(proc_functional)
        table.update(new_table)
    return table


BUITLINS = build_builtin_tables()

File: pbhhg_js/interpret.js
import * as AS from './abstractSyntax.js'
import * as E from './error.js'
import { encodeNumber } from './parse.js'
import {
  checkArity,
  checkType,
  chooseConstructorLike,
  isLiteralExpr,
} from './utils.js'

import buildArithmetics from './builtins/arithmetics.js'
import buildConstructors from './builtins/constructors.js'
import buildFunctional from './builtins/functional.js'
import buildIO from './builtins/io.js'
import buildLogic from './builtins/logic.js'
import buildModule from './builtins/module.js'
import buildSequence from './builtins/sequence.js'
import buildString from './builtins/string.js'

const BUILTINS = [
  buildArithmetics,
  buildConstructors,
  buildFunctional,
  buildIO,
  buildLogic,
  buildModule,
  buildSequence,
  buildString,
].reduce(
  (acc, builder) => Object.assign(acc, builder(procFunctional, strict)),
  {}
)

/**
 * Finds the built-in function corresponding to the builtin function id.
 * @param {AS.Metadata} metadata Caller's metadata.
 * @param {Number} id Built-in function ID.
 * @returns {Function} corresponding function that takes arguments.
 */
function findBuiltin(metadata, id) {
  const inst = encodeNumber(id)
  const builtinFun = BUILTINS[inst]
  if (builtinFun) return builtinFun
  throw new E.UnsuspectedHangeulNotFoundError(
    metadata,
    `${inst}라는 이름의 기본 제공 함수를 찾지 못했습니다.`
  )
}

/** Forces strict evaluation of the value */
export function strict(value) {
  if (value instanceof AS.ExprV) {
    if (value.cache) {
      if (value.cache instanceof AS.UnsuspectedHangeulError) {
        throw value.cache
      }
      return value.cache
    }
    try {
      value.cache = strict(interpret(value.expr, value.env))
      return value.cache
    } catch (error) {
      if (error instanceof AS.UnsuspectedHangeulError) {
        value.cache = error
      }
      throw error
    }
  }
  return value
}

function _accessArray(arr, rel) {
  if (rel >= 0) return arr[rel]
  else return arr[arr.length + rel]
}
function _accessBuffer(buf, rel) {
  if (rel >= buf.byteLength) return undefined
  if (rel < 0) rel += buf.byteLength
  return buf.slice(rel, rel + 1)
}

/** Evaluates the expression in given environment and returns a value */
export function interpret(expr, env) {
  if (expr instanceof AS.Literal) {
    return new AS.IntegerV(expr.value)
  } else if (expr instanceof AS.FunRef) {
    const fun = _accessArray(env.funs, -expr.rel - 1)
    if (fun == null) {
      throw new E.UnsuspectedHangeulOutOfRangeError(
        expr.metadata,
        '함수 참조의 범위를 벗어났습니다.'
      )
    }
    return fun
  } else if (expr instanceof AS.ArgRef) {
    if (env.funs.length !== env.args.length) {
      throw Error(
        'Assertion Error: Environment has ' +
          env.funs.length +
          ' funs and ' +
          env.args.length +
          ' args.'
      )
    }
    var args = _accessArray(env.args, -expr.relF - 1)
    if (args == null) {
      throw new E.UnsuspectedHangeulOutOfRangeError(
        expr.metadata,
        '존재하지 않는 함수에 대한 인수 참조를 시도했습니다.'
      )
    }
    var relA = strict(interpret(expr.relA, env))
    checkType(relA, AS.IntegerV)
    relA = relA.value
    if (relA.lt(0) || relA.geq(args.length)) {
      throw new E.UnsuspectedHangeulOutOfRangeError(
        expr.metadata,
        `${relA}번째 인수를 참조하는데 ${args.length}개의 인수만 받았습니다.`
      )
    } else return args[relA]
  } else if (expr instanceof AS.FunDef) {
    var newFuns = env.funs.slice()
    var newEnv = new AS.Env(newFuns, env.args, env.utils)
    var closure = new AS.ClosureV(expr.body, newEnv)
    newEnv.funs.push(closure)
    return closure
  } else if (expr instanceof AS.FunCall) {
    var fun = new AS.ExprV(expr.fun, env, null)
    var argv = expr.argv.map(function (arg) {
      return new AS.ExprV(arg, env, null)
    })
    var recipe = procFunctional(expr.metadata, fun, AS.CallableV)
    return recipe(argv)
  }
  throw Error('Unexpected expression: ' + expr)
}

/**
 * Converts function-like values into functions.
 * @param metadata Caller's metadata.
 * @param fun A maybe-Expr value that may correspond to a function.
 * @param allow A list of types that are allowed for execution.
 * @returns A recipe function that receives argument list and returns the value.
 */
export function procFunctional(metadata, fun, allow) {
  if (isLiteralExpr(fun)) return findBuiltin(metadata, fun.expr.value)

  fun = strict(fun)
  allow = allow || []
  allow = allow.concat([AS.FunctionV]) // default
  checkType(fun, allow)

  if (fun instanceof AS.FunctionV) {
    return fun.execute.bind(fun)
  } else if (fun instanceof AS.BooleanV) {
    return function (argv) {
      checkArity(argv, 2)
      return argv[fun.value ? 0 : 1]
    }
  } else if (fun instanceof AS.DictV) {
    return function (argv) {
      checkArity(argv, 1)
      const argKey = strict(argv[0]).asKey(strict)
      const value = fun.value[argKey]
      if (value) return value
      throw new E.UnsuspectedHangeulNotFoundError(
        metadata,
        `사전에 다음 표제가 없습니다: ${argKey}`
      )
    }
  } else if (fun instanceof AS.ComplexV) {
    return function (argv) {
      checkArity(argv, 1)
      const arg = strict(argv[0])
      checkType(arg, AS.IntegerV)
      if (arg.value == 0 || arg.value == 1) {
        return new AS.FloatV(fun.value.toVector()[arg.value])
      }
      throw new E.UnsuspectedHangeulValueError(
        metadata,
        `복소수 객체의 ${arg.value}번째 요소를 접근하려고 했습니다.`
      )
    }
  } else {
    const _access = fun instanceof AS.BytesV ? _accessBuffer : _accessArray
    const length =
      fun instanceof AS.BytesV ? fun.value.byteLength : fun.value.length
    return function (argv) {
      checkArity(argv, 1)
      const arg = strict(argv[0])
      checkType(arg, AS.IntegerV)
      const idx = arg.value
      var item = _access(fun.value, idx)
      if (item == null) {
        throw new E.UnsuspectedHangeulOutOfRangeError(
          expr.metadata,
          `길이 ${length}의 객체의 ${idx}번째 요소를 요청했습니다.`
        )
      }
      const V = chooseConstructorLike(fun, [AS.StringV, AS.BytesV])
      if (V) item = new V(item)
      return item
    }
  }
}
